Index: /home/luigi/FreeBSD/head/sys/dev/e1000/if_lem.c
===================================================================
--- /home/luigi/FreeBSD/head/sys/dev/e1000/if_lem.c	(revision 244673)
+++ /home/luigi/FreeBSD/head/sys/dev/e1000/if_lem.c	(working copy)
@@ -318,6 +318,10 @@
 
 #ifdef DEV_NETMAP	/* see ixgbe.c for details */
 #include <dev/netmap/if_lem_netmap.h>
+uint64_t tsc_irq_start, tsc_irq_end, tsc_irq_delta;
+SYSCTL_DECL(_dev_netmap);
+SYSCTL_UQUAD(_dev_netmap, OID_AUTO, delta,
+		CTLFLAG_RD, &tsc_irq_delta, 0, "");
 #endif /* DEV_NETMAP */
 
 /*********************************************************************
@@ -1335,7 +1339,8 @@
 	struct adapter	*adapter = context;
 	struct ifnet	*ifp = adapter->ifp;
 
-
+	tsc_irq_end = rdtsc();
+	tsc_irq_delta = tsc_irq_end - tsc_irq_start;
 	if (ifp->if_drv_flags & IFF_DRV_RUNNING) {
 		lem_rxeof(adapter, adapter->rx_process_limit, NULL);
 		EM_TX_LOCK(adapter);
@@ -1362,6 +1367,7 @@
 	u32		reg_icr;
 
 	ifp = adapter->ifp;
+	tsc_irq_start = rdtsc();
 
 	reg_icr = E1000_READ_REG(&adapter->hw, E1000_ICR);
 
